Add agent resource

knottnt opened the pull request:
Issue #, if available: 2117

Description of changes:
- Add Agent resource
- Create tags utility file for syncing/retrieving tags
- Add e2e test for Agent resource

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

knottnt committed:
Generate Agent resource

knottnt committed:
Use latest aws-sdk-go-v2 version

- Generate from v1.36.3 aws-sdk-go-v2

- Ignore AdditionalModelRequestFields as it uses an unhandled type

knottnt committed:
Update generator.yaml

Add comment for why AdditionalModelRequestFields is ignored.

knottnt committed:
Re-gen with StorageDays marked as a bad default field.

knottnt committed:
Update .gitignore

ignore .vscode

knottnt committed:
Get Agent Creation and Deletion working

- Mark agentResourceRoleARN as required
- Add hooks for calling PrepareAgent
- Ignore ClientToken
- Setup initial Terminal Codes
- Wait for PREPARED state for sync success

knottnt committed:
Call PrepareAgent after CREATE

- Add virtual Spec.AgentStatus field in delta comparison to allow update to be triggered when Agent not in Prepared status

- Add hooks to call PrepareAgent in sdkCreate and sdkUpdate

knottnt committed:
Only update non-default PromptConfigurations

- Add sdkUpdate hook to filter out default PromptConfigurations from the UpdateAgent request

- Add hook to only compare non-default PromptConfigurations

- Add Support for tagging resources

knottnt committed:
Add e2e test for Agent resource

knottnt commented:
/test all

knottnt committed:
Update __init__.py

Fix bad code-gen name

knottnt commented:
/test bedrockagent-kind-e2e


<br class="Apple-interchange-newline">

knottnt committed:
Update conftest.py

fix bad code-gen name

knottnt commented:
/test bedrockagent-kind-e2e

knottnt committed:
Update test_agent.py

knottnt commented:
/test bedrockagent-kind-e2e

knottnt committed:
Create sync_test.go

Add unit tests for tag sync functions

knottnt committed:
Update __init__.py

Fix bad code-gen name in CRD_GROUP

knottnt commented:
/test bedrockagent-kind-e2e
/test bedrockagent-unit-test

knottnt committed:
Update test_agent.py

fix validation error with test agent instruction

knottnt commented:
/test bedrockagent-kind-e2e

knottnt committed:
Update test_agent.py

- Fix bad inference profile region validation error.

knottnt commented:
/test bedrockagent-kind-e2e

knottnt committed:
Update test_agent.py

knottnt commented:
/test bedrockagent-kind-e2e

knottnt committed:
Update agent.py

knottnt commented:
/test bedrockagent-kind-e2e

knottnt committed:
Update agent.py

use keyword arg for boto3 call

knottnt commented:
/test bedrockagent-kind-e2e

knottnt committed:
Re-run code-gen with latest changes.

a-hilaly commented on generator.yaml:
```diff
@@ -12,3 +12,45 @@ ignore:
+    exceptions:
+      terminal_codes:
+        - ValidationException
+        - ResourceNotFoundException
```
resource not found is terminal for bedorck?

knottnt replied:
Hmm for some reason I had thought ResourceNotFound was returned if you provided an invalid arn, but that actually results in an AccessDenied. Given, that AccessDenied could just mean a policy needs to updated I'll just remove this terminal code.

a-hilaly commented on apis/v1alpha1/generator.yaml:
```diff
@@ -12,3 +12,45 @@ ignore:
+        # AgentResourceRoleARN is not marked as required in CreateAgent, but is required by UpdateAgent
+        is_required: true
+      PromptOverrideConfiguration:
+        late_initialize: {}
```
🤞

knottnt replied:
Any, alternatives that you feel would be safer for reading defaults after the initial Create? Definitely, open to changing this.

michaelhtm replied:
late_initialize is just an extra readOne call. I'd say it should be okay to use it?

a-hilaly replied:
i'm just surprised to that this works, i just avoided it in general as I found it usually unnecessary. But hey it's cool if it works :)

knottnt replied:
@a-hilaly How do you handle AWS defaults that are set after the initial create without using late_initialize?

michaelhtm replied:
i'd say calling ReadOne again is the way to go for this one

a-hilaly commented on templates/hooks/agent/sdk_read_one_post_set_output.go.tpl:
```diff
@@ -0,0 +1 @@
+    ko.Spec.Tags, err = rm.getTags(ctx, string(*ko.Status.ACKResourceMetadata.ARN))
```
missing error handling

knottnt replied:
`err` is now returned if it is set.

a-hilaly commented on templates/hooks/agent/delta_pre_compare.go.tpl:
```diff
@@ -0,0 +1,5 @@
+    // Hack to ensure that reconcile loop triggers update for PrepareAgent call
+	// if AgentStatus is not in PREPARED state.
+	compareAgentStatus(b.ko.Status.AgentStatus, delta)
+
+	comparePropertyOverrideConfiguration(a, b, delta)
```
super tiny nit: generally i try to use delta as first parameter

knottnt replied:
re-ordered the function parameters.

a-hilaly commented on templates/hooks/agent/sdk_create_post_set_output.go.tpl:
```diff
@@ -0,0 +1,6 @@
+    // Agent is created in NOT_PREPARED state which is not fully ready.
+    // To complete setup need to call PrepareAgent. Note this call may
+    // fail as Agent has been created recently. Will need to ensure that 
+    // subsequent reconcile loops retry if Agent is not PREPARED.
+    prepareAgent(ctx, rm.sdkapi, rm.metrics, *ko.Status.AgentID)
```
shall we just requeue here and delegate the preparation to the update call? just making the flow consistent

knottnt replied:
It looks like the controller will requeue due to AgentStatus.PREPARED being set as prerequisite for being synced. So, I just removed this hook.

a-hilaly commented on pkg/resource/tags/sync_test.go:
Nice!

a-hilaly reviewed (commented):
Really good stuff, thanks Trevor! left a few comments inline

knottnt committed:
PR fixes

- Move delta parameter for customer compare methods

- Add error handing to read_one_post_set_output hook

- Remove prepareAgent sdk_create_post_set_output hook

- Remove ResourceNotFound as terminal code

michaelhtm commented on pkg/resource/agent/manager.go:
```diff
@@ -0,0 +1,421 @@
+	observed acktypes.AWSResource,
+	latest acktypes.AWSResource,
+) acktypes.AWSResource {
+	observedKo := rm.concreteResource(observed).ko.DeepCopy()
+	latestKo := rm.concreteResource(latest).ko.DeepCopy()
+	if observedKo.Spec.PromptOverrideConfiguration != nil && latestKo.Spec.PromptOverrideConfiguration == nil {
+		latestKo.Spec.PromptOverrideConfiguration = observedKo.Spec.PromptOverrideConfiguration
+	}
```
seems like these may require you to add an sdkFind hook to set latestKo.Spec.PromptOverrideConfiguration = nil for this late_initialize to succeed..

knottnt replied:
This is just the standard generated late_initialize code. It does seem that if the field somehow got set before late initialization it isn't necessarily a problem. If PromptConfigurations was already set wouldn't that just mean that the Create operation had returned a value for it?

michaelhtm replied:
Yes that is right!

knottnt replied:
Okay, I think it should be fine to leave it as-is then. If they happen to return the default in the Create operation's response we should be okay using those.

michaelhtm commented on templates/hooks/agent/sdk_update_pre_build_request.go.tpl:
```diff
@@ -0,0 +1,21 @@
+    // If AgentStatus is not in PREPARED state we need to call PrepareAgent to finalize setup
+    // of Agent. Uses hack (see delta.go) to trigger update from non-existent Spec.AgentStatus
+    if delta.DifferentAt("Spec.AgentStatus") {
+		prepareAgent(ctx, rm.sdkapi, rm.metrics, *desired.ko.Status.AgentID)
```
should we check error here?

michaelhtm replied:
and set AgentStatus to Prepared if successful?

michaelhtm replied:
Or use the PrepareAgent response to set `latest.ko.Status`?

knottnt replied:
Right now I just leave it for the next requeue to sync the new AgentStatus. Are there any issues with doing that?

knottnt replied:
As for returning the error, yes that's a good catch! Definitely a bit of an adjustment coming from the C# and Kotlin/Java world.

knottnt replied:
`err` is now returned when it is set.

michaelhtm commented on test/e2e/resources/agent.yaml:
```diff
@@ -0,0 +1,49 @@
+  description: $AGENT_DESCRIPTION
+  instruction: $AGENT_INSTRUCTION
+  foundationModel: $AGENT_MODEL
+  agentResourceRoleARN: $AGENT_ROLE_ARN
```
should we add a reference for Role here?

knottnt replied:
Oh yeah would a be a good thing to add. Is it just a generator.yaml config setting?

knottnt replied:
Added reference to IAM Role resource.

michaelhtm commented on test/e2e/tests/test_agent.py:
```diff
@@ -0,0 +1,137 @@
+@service_marker
+@pytest.mark.canary
+class TestAgent:
+    def test_crud(self, simple_agent):
```
should we also add tag tests?

knottnt replied:
Test for updating tags has been added.

michaelhtm reviewed (commented):
Thanks @knottnt! left a few comments

a-hilaly reviewed (approved):
Neat! Thanks a lot Trevor!

knottnt committed:
PR changes

- Add test for updating Agent tags
- Return error from prepareAgent call in sdkUpdate
- Add reference to IAM role for agentResourceRoleARN

michaelhtm reviewed (approved):
Nice 👍

michaelhtm commented:
/lgtm
